From: Amir Vadai <amirva@mellanox.com>
Date: Fri, 13 May 2016 12:55:36 +0000
Subject: net/sched: act_gact: Update statistics when offloaded to hardware
Patch-mainline: v4.7-rc1
Git-commit: 9fea47d93bcc98946a6eca0f019ced337564a344
References: bsc#1015342 FATE#321688 bsc#1015343 FATE#321689

Implement the stats_update callback that will be called by NIC drivers
for hardware offloaded filters.

Signed-off-by: Amir Vadai <amirva@mellanox.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Benjamin Poirier <bpoirier@suse.com>
---
 net/sched/act_gact.c |   15 +++++++++++++++
 1 file changed, 15 insertions(+)

--- a/net/sched/act_gact.c
+++ b/net/sched/act_gact.c
@@ -145,6 +145,20 @@ static int tcf_gact(struct sk_buff *skb,
 	return action;
 }
 
+static void tcf_gact_stats_update(struct tc_action *a, u64 bytes, u32 packets,
+				  u64 lastuse)
+{
+	struct tcf_gact *gact = a->priv;
+	int action = READ_ONCE(gact->tcf_action);
+	struct tcf_t *tm = &gact->tcf_tm;
+
+	_bstats_cpu_update(this_cpu_ptr(gact->common.cpu_bstats), bytes, packets);
+	if (action == TC_ACT_SHOT)
+		this_cpu_ptr(gact->common.cpu_qstats)->drops += packets;
+
+	tm->lastuse = lastuse;
+}
+
 static int tcf_gact_dump(struct sk_buff *skb, struct tc_action *a, int bind, int ref)
 {
 	unsigned char *b = skb_tail_pointer(skb);
@@ -188,6 +202,7 @@ static struct tc_action_ops act_gact_ops
 	.type		=	TCA_ACT_GACT,
 	.owner		=	THIS_MODULE,
 	.act		=	tcf_gact,
+	.stats_update	=	tcf_gact_stats_update,
 	.dump		=	tcf_gact_dump,
 	.init		=	tcf_gact_init,
 };
